src/liballoc/alloc.rs

```diff
@@ -191,12 +193,78 @@ pub(crate) unsafe fn box_free<T: ?Sized>(ptr: Unique<T>) {
 #[stable(feature = "global_alloc", since = "1.28.0")]
 #[rustc_allocator_nounwind]
 pub fn oom(layout: Layout) -> ! {
-    #[allow(improper_ctypes)]
-    extern "Rust" {
-        #[lang = "oom"]
-        fn oom_impl(layout: Layout) -> !;
+    let hook = HOOK.load(Ordering::SeqCst);
+    let hook: fn(Layout) = if hook.is_null() {
+        default_oom_hook
+    } else {
+        unsafe { mem::transmute(hook) }
+    };
+    hook(layout);
+    unsafe {
+        #[cfg(not(stage0))]
+        {
+            #[allow(improper_ctypes)]
+            extern "Rust" {
+                // This function is generated by the compiler
+                // and calls __rust_abort_internal in src/libstd/alloc.rs
+                // if the `std` crate is linked.
+                fn __rust_maybe_abort_internal();
+            }
+            __rust_maybe_abort_internal();
+        }
+        // The above does nothing and returns when `std` is not linked.
+        // In that case we abort "naively" here:
+        ::core::intrinsics::abort()
+    }
+}
+
+fn default_oom_hook(layout: Layout) {
+    #[cfg(stage0)]
+    let _ = layout;
+    #[cfg(not(stage0))]
+    unsafe {
+        #[allow(improper_ctypes)]
+        extern "Rust" {
+            // This function is generated by the compiler
+            // and calls __rust_default_oom_hook in src/libstd/alloc.rs
+            // if the `std` crate is linked.
+            fn __rust_maybe_default_oom_hook(size: usize, align: usize);
+        }
+        __rust_maybe_default_oom_hook(layout.size(), layout.align())
+    }
+}
+
+static HOOK: AtomicPtr<()> = AtomicPtr::new(ptr::null_mut());
+
+/// Registers a custom OOM hook, replacing any that was previously registered.
+///
+/// The OOM hook is invoked when an infallible memory allocation fails, before
+/// the runtime aborts. The default hook prints a message to standard error,
+/// but this behavior can be customized with the [`set_oom_hook`] and
+/// [`take_oom_hook`] functions.
+///
+/// The hook is provided with a `Layout` struct which contains information
+/// about the allocation that failed.
+///
+/// The OOM hook is a global resource.
+#[unstable(feature = "oom_hook", issue = "51245")]
+pub fn set_oom_hook(hook: fn(Layout)) {
+    HOOK.store(hook as *mut (), Ordering::SeqCst);
+}
+
+/// Unregisters the current OOM hook, returning it.
+///
+/// *See also the function [`set_oom_hook`].*
+///
+/// If no custom hook is registered, the default hook will be returned.
+#[unstable(feature = "oom_hook", issue = "51245")]
+pub fn take_oom_hook() -> fn(Layout) {
+    let hook = HOOK.swap(ptr::null_mut(), Ordering::SeqCst);
+    if hook.is_null() {
+        default_oom_hook
```

Comment:
This does not "take" the default hook, though. Should be reflected in the documentation.

Reply:
Would you expect that "taking" the default would replace it with a no-op? This is a behavior change not directly related to this PR and is maybe better discussed on the tracking issue: https://github.com/rust-lang/rust/issues/51245